File: backend/src/main/java/com/coing/global/exception/RestExceptionHandler.java
package com.coing.global.exception;

import java.util.stream.Collectors;

import org.springframework.context.MessageSource;
import org.springframework.context.i18n.LocaleContextHolder;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import com.coing.util.BasicResponse;

import io.jsonwebtoken.ExpiredJwtException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@RestControllerAdvice
@Slf4j
@RequiredArgsConstructor
public class RestExceptionHandler {

	private final MessageSource messageSource;

	@ExceptionHandler(BusinessException.class)
	public ResponseEntity<BasicResponse> handleBusinessException(BusinessException exception) {
		log.warn("[ExceptionHandler] Message: {}, Detail: {}", exception.getMessage(), exception.getDetail());
		if (exception.getMessage() != null && exception.getMessage().contains("인증 토큰")) {
			return ResponseEntity.status(HttpStatus.FORBIDDEN).body(BasicResponse.of(exception));
		}
		return BasicResponse.to(exception);
	}

	@ExceptionHandler(ExpiredJwtException.class)
	public ResponseEntity<BasicResponse> handleExpiredJwtException(ExpiredJwtException ex) {
		log.warn("[ExceptionHandler] ExpiredJwtException: {}", ex.getMessage());
		BusinessException businessException = new BusinessException("토큰 만료", HttpStatus.FORBIDDEN, ex.getMessage());
		return ResponseEntity.status(HttpStatus.FORBIDDEN).body(BasicResponse.of(businessException));
	}

	@ExceptionHandler(RuntimeException.class)
	public ResponseEntity<BasicResponse> handleRuntimeException(RuntimeException exception) {
		log.error("[ExceptionHandler] Runtime exception occurred: ", exception);

		HttpStatus status = HttpStatus.INTERNAL_SERVER_ERROR;

		BusinessException businessException = new BusinessException("Internal Server Error",
			HttpStatus.INTERNAL_SERVER_ERROR, exception.getMessage());
		log.error("[ExceptionHandler] Message: {}, Detail: {}", businessException.getMessage(),
			businessException.getDetail());

		return ResponseEntity
			.status(status)
			.body(BasicResponse.of(businessException));
	}

	@ExceptionHandler(MethodArgumentNotValidException.class)
	public ResponseEntity<BasicResponse> handleValidationExceptions(MethodArgumentNotValidException ex) {
		String errors = ex.getBindingResult().getFieldErrors().stream()
			.map(error -> messageSource.getMessage(error, LocaleContextHolder.getLocale()))
			.collect(Collectors.joining(", "));
		BusinessException businessException = new BusinessException("Validation Error", HttpStatus.BAD_REQUEST, errors);
		log.warn("[ExceptionHandler] Validation errors: {}", errors);
		return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(BasicResponse.of(businessException));
	}
}

File: backend/src/main/java/com/coing/global/exception/RestExceptionHandler.kt
package com.coing.global.exception

import com.coing.util.BasicResponse
import io.jsonwebtoken.ExpiredJwtException
import org.springframework.context.MessageSource
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.slf4j.LoggerFactory

@RestControllerAdvice
class RestExceptionHandler(
	private val messageSource: MessageSource
) {

	private val log = LoggerFactory.getLogger(RestExceptionHandler::class.java)

	@ExceptionHandler(BusinessException::class)
	fun handleBusinessException(exception: BusinessException): ResponseEntity<BasicResponse> {
		log.warn("[ExceptionHandler] Message: {}, Detail: {}", exception.message, exception.detail)
		return if (exception.message?.contains("인증 토큰") == true) {
			ResponseEntity.status(HttpStatus.FORBIDDEN).body(BasicResponse.of(exception))
		} else {
			BasicResponse.to(exception)
		}
	}

	@ExceptionHandler(ExpiredJwtException::class)
	fun handleExpiredJwtException(ex: ExpiredJwtException): ResponseEntity<BasicResponse> {
		log.warn("[ExceptionHandler] ExpiredJwtException: {}", ex.message)
		val businessException = BusinessException("토큰 만료", HttpStatus.FORBIDDEN, ex.message ?: "")
		return ResponseEntity.status(HttpStatus.FORBIDDEN).body(BasicResponse.of(businessException))
	}

	@ExceptionHandler(RuntimeException::class)
	fun handleRuntimeException(exception: RuntimeException): ResponseEntity<BasicResponse> {
		log.error("[ExceptionHandler] Runtime exception occurred: ", exception)

		val businessException = BusinessException("Internal Server Error", HttpStatus.INTERNAL_SERVER_ERROR, exception.message ?: "")
		log.error("[ExceptionHandler] Message: {}, Detail: {}", businessException.message, businessException.detail)

		return ResponseEntity
			.status(HttpStatus.INTERNAL_SERVER_ERROR)
			.body(BasicResponse.of(businessException))
	}

	@ExceptionHandler(MethodArgumentNotValidException::class)
	fun handleValidationExceptions(ex: MethodArgumentNotValidException): ResponseEntity<BasicResponse> {
		val errors = ex.bindingResult.fieldErrors
			.joinToString(", ") { error ->
				messageSource.getMessage(error, LocaleContextHolder.getLocale())
			}

		val businessException = BusinessException("Validation Error", HttpStatus.BAD_REQUEST, errors)
		log.warn("[ExceptionHandler] Validation errors: {}", errors)
		return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(BasicResponse.of(businessException))
	}
}
